### PR TITLE
Expose the alias of a network interface

### PR DESCRIPTION
**Pull request by dornand**
`NetworkIF.getIfAlias()` for Windows and Linux.  Relates to #1541

### CODE REVIEW

**Comment by dbwiddis** [comment]
Note to self: for macOS, looks like [SCNetworkInterfaceCopyAll](https://developer.apple.com/documentation/systemconfiguration/1517090-scnetworkinterfacecopyall?language=objc) can enumerate interfaces. Not sure whether any of the other commands include the alias though.

**Comment by dbwiddis** [comment]
Looking more into this, what is the difference between the Alias and the "DisplayName" we already use?

**Comment by dornand** [comment]
On Windows i get:
Name: eth1
Display name: Intel(R) Ethernet Connection (4) I219-V
Alias: Ethernet

`ipconfig` is using the alias here:
Ethernet adapter _Ethernet_:

   Connection-specific DNS Suffix  . : 
   Link-local IPv6 Address . . . . . : fe80::2916...
   IPv4 Address. . . . . . . . . . . : 10.11...
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : 10.11...

Also the network adapter is displayed as "Ethernet" in the control center.

**Comment by dornand** [comment]
Actually on Windows `NetworkInterface.getDisplayName()` is returning the device description (=MIB_IF_ROW2.Description), instead of the "real" display name, which is the alias.

**Comment by dornand** [comment]
In general though, the alias is the custom name, given to a network interface by an administrator.
So `NetworkIF.getAlias()` could be useful, where such custom names exist.

**Comment by dbwiddis** [comment]
Thanks for the details.  There's probably a default version we can produce parsing the output of `ifconfig` on *nix machines.  I'll research more tonight.

**Comment by dbwiddis** [comment]
Interestingly, looking at the JDK source code, the [description](https://github.com/openjdk/jdk/blob/master/src/java.base/windows/native/libnet/NetworkInterface.c#L265) is fetched, probably because they chose the older [MIB_IFROW](https://docs.microsoft.com/en-us/windows/win32/api/ifmib/ns-ifmib-mib_ifrow) structure for compatibility.  All the other implementations [just use the same value for name and displayName](https://github.com/openjdk/jdk/blob/master/src/java.base/unix/native/libnet/NetworkInterface.c#L719).

I'm still confused, though, it seems you can assign [multiple aliases](https://linuxtechlab.com/ip-aliasing-multiple-ip-single-nic/) to a NIC. How do we handle that (or how does `sysfs` handle it)?  Or is the `IfAlias` we're mapping completely different from an "IP Alias"?  In which case perhaps it should be named `getIfAlias()`.

File: oshi-core/src/main/java/oshi/hardware/NetworkIF.java [hunk 1]
```diff
@@ -61,6 +61,13 @@
      */
     String getDisplayName();
 
+    /**
+     * Interface alias.
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/NetworkIF.java:65]: Let's me more verbose and describe the IfAlias per the RFC so it's clear what we're talking about.

File: oshi-core/src/main/java/oshi/hardware/NetworkIF.java [hunk 2]
```diff
@@ -61,6 +61,13 @@
...
+     *
+     * @return The alias of the network interface.
+     */
+    String getAlias();
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/NetworkIF.java:69]: I think `getIfAlias()` will make it more clear what this means.

File: oshi-core/src/main/java/oshi/hardware/common/AbstractNetworkIF.java [hunk 3]
```diff
@@ -229,6 +229,12 @@ public boolean isConnectorPresent() {
         return false;
     }
 
+    @Override
+    public String getAlias() {
+        // default
+        return "";
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/common/AbstractNetworkIF.java:235]: Is there a default parsing of `ifconfig` we can put in here that's reasonably consistent on *nix machines?

File: oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsNetworkIF.java [hunk 4]
```diff
@@ -179,6 +185,7 @@ public boolean updateAttributes() {
             this.collisions = ifRow.OutDiscards; // closest proxy
             this.inDrops = ifRow.InDiscards; // closest proxy
             this.speed = ifRow.ReceiveLinkSpeed;
+            this.alias = new String(ifRow.Alias).trim();
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsNetworkIF.java:188]: Use JNA's `Native.toString(char[])` here.

**Review by dbwiddis — COMMENTED** [review]
Some suggestions while I research the other platforms.

**Comment by dornand** [comment]
I think IP Aliasing is a completely different thing. That's why we have `String[] getIPv4addr()`, to map several IP addresses.

**Comment by dbwiddis** [comment]
Yeah. I've spent way too much time going down rabbit holes on this.  My observation:
 - Windows has 3 string fields, the name, alias, and description
 - The JDK uses the description for the "displayName" and ignores the alias (for compatibility) on Windows
 - The JDK simply duplicates the name into displayName for all other OS's
 - macOS does have a "LocalizedDisplayName" accessible in the System Configuration framework which isn't really an ifAlias. Would be nice to override the display name for macOS though.
 - Other than that there's no easy/consistent/fast way to get to it.

I'll see if I can work up the macOS change to the display name tomorrow, other than that this is probably as good as we're going to get.  Probably want to tweak the toString.

**Comment by dbwiddis** [comment]
Oh, is there a "description" field in Linux we can put in the displayName field?

**Comment by dornand** [comment]
i'll have to check later. could not find "ifDescr" exposed in sysfs.

i found a discussion suggesting to use the "pci_lookup_name" for ifDescr:
[https://sourceforge.net/p/net-snmp/mailman/message/26092573/](https://sourceforge.net/p/net-snmp/mailman/message/26092573/)

**Comment by dbwiddis** [comment]
That's probably overkill.  Another option is `lspci` for example
```
00:05.0 Ethernet controller: Red Hat, Inc. Virtio network device
```
PCI bus 0, Slot 5 `enp0s5`: `/sys/class/net/enp0s5`

But I don't like command lines when I can avoid them.

**Comment by dbwiddis** [comment]
I think I may have a native solution for the name using udev.  
```
$ udevadm info /sys/class/net/enp0s5
...
E: ID_MODEL_FROM_DATABASE=Virtio network device
E: ID_MODEL_ID=0x1000
...
E: ID_VENDOR_FROM_DATABASE=Red Hat, Inc.
E: ID_VENDOR_ID=0x1af4
...
```
However that may just be the same "pci_lookup_name" lookup you described earlier from the model and vendor ID which I expect are keys, while the DB lookup is probably done internally in `udevadm` and not stored...

**Comment by dbwiddis** [comment]
See code in `LinuxHWDiskStore` starting at line 167 as an example, but use the sysfs path and try the `ID_MODEL_FROM_DATABASE` property value above.

**Comment by dbwiddis** [comment]
I'm good to merge this unless you think there is a need for any other changes.  I'd appreciate you testing on your own Linux as well.  It was pretty trivial to grab the vendor and model strings from udev, simulating the `pci_lookup_name` approach.

File: oshi-core/src/main/java/oshi/hardware/platform/linux/LinuxNetworkIF.java [hunk 5]
```diff
@@ -162,6 +194,7 @@ public boolean updateAttributes() {
         String collisionsPath = String.format("/sys/class/net/%s/statistics/collisions", getName());
         String rxDropsPath = String.format("/sys/class/net/%s/statistics/rx_dropped", getName());
         String ifSpeed = String.format("/sys/class/net/%s/speed", getName());
+        String ifAlias = String.format("/sys/class/net/%s/ifalias", getName());
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/platform/linux/LinuxNetworkIF.java:197]: CI complains this local variable name shadows the class field name.  Must be a popular name. :)

**Comment by dornand** [comment]
Ah, now we need to test those default methods separately.
Should I just make a NetworkIFTest?

**Comment by dbwiddis** [comment]
Not sure what you mean about the testing? Should give identical results to the previous version?   (We mostly just test the API... yes I know that skips a lot... we'll do it right on the redesign from scratch! :grin: )

**Comment by dornand** [comment]
Just noticed the failing coverage check  :)

**Comment by dbwiddis** [comment]
Yeah it's annoying but I can't find a way to set a threshold. It counts comments too!  We can ignore it -- you added the necessary test for the new API.

**Comment by dbwiddis** [comment]
Coverage is really only reported on the linux branch but we run all the tests on all 7 supported OS's. 

So are we good to merge this? Anything else you need to add?

**Comment by dornand** [comment]
I'm ok to merge. Thanks for the support!

**Pull request merged**